packages/agoric-cli/src/lib/bundles.js

```diff
@@ -0,0 +1,88 @@
+/* global Buffer */
+import { BlobReader, TextWriter, ZipReader } from '@zip.js/zip.js';
+import fs from 'node:fs';
+import path from 'node:path';
+
+/** @import {Bundle} from '@agoric/swingset-vat'; */
+
+/**
+ * @typedef CompartmentMap
+ * @property {string[]} tags
+ * @property {{compartment: string, module: string}} entry
+ * @property {Record<string, {name: string, label: string, location: string, modules: Record<string, {compartment: string, module: string}>}} compartments
+ */
+
+/** @param {Bundle} bundleObj*/
+export const extractBundleInfo = async bundleObj => {
+  await null;
+  if (bundleObj.moduleFormat === 'endoZipBase64') {
+    const contents = Buffer.from(bundleObj.endoZipBase64, 'base64');
+    const zipBlob = new Blob([contents], { type: 'application/zip' });
+
+    const zipFileReader = new BlobReader(zipBlob);
+    const writer = new TextWriter('utf-8');
+
+    const zipReader = new ZipReader(zipFileReader);
+    const entries = await zipReader.getEntries();
+    await zipReader.close();
+
+    assert(entries[0].filename, 'compartment-map.json');
+    const cmapStr = await entries[0].getData(writer);
+    /** @type {CompartmentMap} */
+    const compartmentMap = JSON.parse(cmapStr);
+
+    const fileSizes = Object.fromEntries(
+      entries.map(e => [e.filename, e.compressedSize]),
+    );
+
+    return { compartmentMap, fileSizes };
+  }
+};
+
+// UNTIL https://github.com/endojs/endo/issues/1656
+/** @param {string} bundleFilename */
+export const statBundle = async bundleFilename => {
+  const bundle = fs.readFileSync(bundleFilename, 'utf8');
+  /** @type {Bundle} */
+  const bundleObj = JSON.parse(bundle);
+  console.log('\nBUNDLE', bundleObj.moduleFormat, bundleFilename);
+
+  const info = await extractBundleInfo(bundleObj);
+
+  // XXX use objectMap
+  const byCompartment = Object.fromEntries(
+    Object.entries(info.compartmentMap.compartments).map(([k, v]) => {
+      let sum = 0;
+      for (const m of Object.values(v.modules)) {
+        // XXX surely Endo has some code for this
```

Comment:
There is code that traverses a compartment map, but not a traversal abstraction. I think I prefer this clean and readable loop structure over a traversal abstraction that obscures the structure of the compartment map schema. The schema is the API.